rust/core-lib/src/event_context.rs

```diff
@@ -494,6 +495,25 @@ impl<'a> EventContext<'a> {
         self.editor.borrow_mut().dec_revs_in_flight();
     }
 
+    /// Checks a newline needs to be added to the end of the file
+    pub(crate) fn pre_save_newline(&mut self) -> bool {
+        let editor = self.editor.borrow();
+        let ed_text = editor.get_buffer();
+        let view = self.view.borrow();
+        let offset = view.line_of_offset(ed_text, ed_text.len()) + 1;
+        if let Some(last_char) = ed_text
+            .lines_raw(Interval::new(offset - 1, view.offset_of_line(ed_text, offset)))
```

Comment:
I don't think this is going to work as expected. From the docs for `Rope::lines_raw`:

> An iterator over the raw lines. The lines, except the last, include the
> terminating newline.

So assuming the docs are correct, this will never return the trailing newline. Since all we really need to do is check the last byte, we should be able to write,

```rust
let rope_len = rope.len();
if rope_len > 0 && rope.byte_at(rope_len - 1) == b'\n' {
// .. logic
}
```

Reply:
> An iterator over the raw lines. The lines, except the last, include the
terminating newline.

Ah, it does return the last line with a newline, it just assumes that that naturally never happens :)

Your suggestion is way prettier though